File: examples/solv_module_ts.rb
#
# jruby examples for jas.
# $Id$
#

require "examples/jas"

# Ore extension solvable polynomial example, modules

rp = PolyRing.new(QQ(),"x,y,z,t",PolyRing.lex);
#is automatic: [one,x,y,z,t] = rp.gens();

trel = [ z, y,  y * z + x,
         t, y,  y * t + y,
         t, z,  z * t - z
       ];

puts "trel: = [" + trel.map { |r| r.to_s }.join(", ") + "]";
puts;

rs = SolvPolyRing.new(QQ(),"x,y,z,t",PolyRing.lex,trel);

#exit(0)

f = rs.ideal("",[t**2 + z**2 + x**2 + y**2 + 1]);
puts "f: " + str(f);
tf = f.twosidedGB();
puts "t: " + str(tf);
puts;

#exit(0)

r = SolvableModule.new("",rs);
puts "SolvableModule: " + str(r);
puts;

subm = [
         [ 0, t**2 + z**2 + x**2 + y**2 + 1],
         [ x**2 + y**2,     z ]
       ];

m = SolvableSubModule.new( r, "", subm );
puts "SolvableSubModule: " + str(m);
puts;

#exit()
#startLog();

lg = m.leftGB();
puts "seq left GB: " + str(lg);
puts "is left GB: " + str(lg.isLeftGB());
puts;

tg = m.twosidedGB();
puts "seq twosided GB: " + str(tg);
puts "is twosided GB: " + str(tg.isTwosidedGB());
puts "is right GB: " + str(tg.isRightGB());
puts;

#exit()

rg = m.rightGB();
puts "seq right GB: " + str(rg);
puts "is right GB: " + str(rg.isRightGB());
puts;


# as quotients to coefficients
rq = SRF(rs);

rpq = PolyRing.new(rq,"v,w",PolyRing.lex);
puts "PolyRing: rpq = " + str(rpq);

vrel = [ v, t, t * v + x,
         w, t, t * w + y
       ];

puts "vrel: = [" + vrel.map { |r| r.to_s }.join(", ") + "]";
puts;

rsq = SolvPolyRing.new(rq,"v,w",PolyRing.lex,vrel);
puts "SolvPolyRing: rsq = " + str(rsq);
puts;


r = SolvableModule.new("",rsq);
puts "SolvableModule: " + str(r);
puts;

subm = [
         [     0, v + 1 ],
         [ w + x, w - v ]
       ];

m = SolvableSubModule.new( r, "", subm );
puts "SolvableSubModule: " + str(m);
puts;

#startLog();

lg = m.leftGB();
puts "seq left GB: " + str(lg);
puts "is left GB: " + str(lg.isLeftGB());
puts;

#exit();
#startLog();

tg = m.twosidedGB();
puts "seq twosided GB: " + str(tg);
puts "is twosided GB: " + str(tg.isTwosidedGB());
puts "is right GB: " + str(tg.isRightGB());
puts;

#exit()

rg = m.rightGB();
puts "seq right GB: " + str(rg);
puts "is right GB: " + str(rg.isRightGB());
puts;


#startLog();
terminate();

File: examples/solv_module_ts.py
#
# jython examples for jas.
# $Id$
#

from java.lang import System

from jas import SolvableRing, SolvPolyRing, PolyRing, SRF 
from jas import QQ, startLog, SolvableModule, SolvableSubModule, terminate

# Ore extension solvable polynomial example, modules

rp = PolyRing(QQ(),"x,y,z,t",PolyRing.lex);
#is automatic: [one,x,y,z,t] = rp.gens();

trel = [ z, y,  y * z + x,
         t, y,  y * t + y,
         t, z,  z * t - z
       ];

print "trel: = " + str([ str(f) for f in trel ]);
print;

rs = SolvPolyRing(QQ(),"x,y,z,t",PolyRing.lex,trel);

#exit(0)

f = rs.ideal("",[t**2 + z**2 + x**2 + y**2 + 1]);
print "f: " + str(f);
tf = f.twosidedGB();
print "t: " + str(tf);
print;

#exit(0)

r = SolvableModule("",rs);
print "SolvableModule: " + str(r);
print;

subm = [
         [ 0, t**2 + z**2 + x**2 + y**2 + 1],
         [ x**2 + y**2,     z ]
       ];

m = SolvableSubModule( r, "", subm );
print "SolvableSubModule: " + str(m);
print;

#exit()
#startLog();

lg = m.leftGB();
print "seq left GB: " + str(lg);
print "is left GB: " + str(lg.isLeftGB());
print;

tg = m.twosidedGB();
print "seq twosided GB: " + str(tg);
print "is twosided GB: " + str(tg.isTwosidedGB());
print "is right GB: " + str(tg.isRightGB());
print;

#exit()

rg = m.rightGB();
print "seq right GB: " + str(rg);
print "is right GB: " + str(rg.isRightGB());
print;


# as quotients to coefficients
rq = SRF(rs);

rpq = PolyRing(rq,"v,w",PolyRing.lex);
print "PolyRing: rpq = " + str(rpq);

vrel = [ v, t, t * v + x,
         w, t, t * w + y
       ];

print "vrel: = " + str([ str(f) for f in vrel ]);
print;

rsq = SolvPolyRing(rq,"v,w",PolyRing.lex,vrel);
print "SolvPolyRing: rsq = " + str(rsq);
print;


r = SolvableModule("",rsq);
print "SolvableModule: " + str(r);
print;

subm = [
         [     0, v + 1 ],
         [ w + x, w - v ]
       ];

m = SolvableSubModule( r, "", subm );
print "SolvableSubModule: " + str(m);
print;

#startLog();

lg = m.leftGB();
print "seq left GB: " + str(lg);
print "is left GB: " + str(lg.isLeftGB());
print;

#exit();
#startLog();

tg = m.twosidedGB();
print "seq twosided GB: " + str(tg);
print "is twosided GB: " + str(tg.isTwosidedGB());
print "is right GB: " + str(tg.isRightGB());
print;

#exit()

rg = m.rightGB();
print "seq right GB: " + str(rg);
print "is right GB: " + str(rg.isRightGB());
print;


#startLog();
terminate();
